### PR TITLE
feat: add Web Test Runner framework and tooling

### PR DESCRIPTION
**Pull request by zhawkins**
Adds [Web Test Runner](https://modern-web.dev/blog/introducing-modern-web/#web-test-runner) which will eventually [Web Component Tester](https://github.com/Polymer/web-component-tester). Yes that means getting rid of Selenium. 😃 

This merge requests adds:

1. The ability to create new tests for either new or existing elements using Yeoman.
2. The ability to run those new tests with web-test-runner.
3. The ability to run tests using the element embedded within a Vue or React wrapper.
4. The ability to run those new tests with Github actions in parallel with existing (legacy) tests.
4. Migrates tests (old and new) to use the CDN versions of React & Vue so we can remove those from project dependencies.

I've converted the tests for `<pfe-select>` as an example.

### Testing instructions

**Setup:**

1. `git checkout poc-web-test-runner`
2. `npm install`
3. `npm run build`

---

#### Verify the new tests run:

**Do this:**

`npm run test:ci`

**You should see:**

Tests should run and pass.

---

#### Verify that new generated elements have new tests

**Do this:**

1. Run `npm run new`
2. Select `container`
3. Use the name `pfe-foo`
4. Select the defaults (hit enter) for everything _except_ "Do you want to use Sass" (pick "n")
5. Run `npm run test:watch`
6. Hit the "F" key to focus a test file

**You should see:**

You should see the new `pfe-foo.spec.js` file as an option.

Hitting "1" should focus that test and they should pass. Exit out of it by using ctrl + c

---

#### Verify that new tests can be added to existing elements

**Do this:**

1. Run `npm run new:test "pfe-tabs"`

**You should see:**

You should see the following file created.

`elements/pfe-tabs/test/pfe-tabs.spec.js`

Open the file and verify it's specific to `pfe-tabs`.

---

#### Verify tests can use React and Vue wrappers

**Do this:**

1. Run `npm run test:watch --group with-react`
2. Once testing starts hit "D" for debug which should open up a browser. (Select pfe-select to focus if it's an option)
3. In dev tools, open up the `<head>` element. Verify React has been added.

**You should see:**

Verify the main `<div>` has a react wrapper around the `pfe-select` element.

**Do this:**

1. Run `npm run test:watch --group with-vue`
2. Once testing starts hit "D" for debug which should open up a browser. (Select pfe-select to focus if it's an option)
3. In dev tools, open up the `<head>` element. Verify Vue has been added.

**You should see:**

Verify the main `<div>` has a Vue wrapper around the `pfe-select` element.

---

### Ready-for-merge Checklist

- [x] Expected files: all files in this pull request are related to one request or issue (no stragglers or scope-creep).
- [x] Repository compiles and tests pass.
- [x] Changelog updated.
- [x] Documentation (README.md, WHY.md, etc.) updated or added.

### Merging

Please **squash** when merging and ensure your commit message uses [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) formatting.

**Be sure to share your updates with the [patternfly-elements-contribute@redhat.com](mailto:patternfly-elements-contribute@redhat.com) mailing list!**

### CODE REVIEW

File: .nvmrc [hunk 1]
```diff
@@ -1 +1 @@
-v12.16.3
+v12.18.2
```
Comment (zhawkins) [inline .nvmrc:1]: Bumped this because we need a higher version of node for esmodule support with Web Test Runner.

File: elements/pfe-select/src/pfe-select.js [hunk 2]
```diff
@@ -80,7 +80,7 @@ class PfeSelect extends PFElement {
     super.disconnectedCallback();
 
     this.observer.disconnect();
-    this._input.removeEventListener("input", this._inputChanged);
+    this._input && this._input.removeEventListener("input", this._inputChanged);
```
Comment (zhawkins) [inline elements/pfe-select/src/pfe-select.js:83]: This fixes a bug where if pfe-select doesn't have any content it breaks JS because there's no `this._input` to remove the event listener on.

File: generators/element/index.js [hunk 3]
```diff
@@ -376,6 +348,15 @@ module.exports = class extends Generator {
         console.log(util.inspect(this.props, { showHidden: false, depth: 4 }));
         console.log("-------------------------------------------//\n");
       }
+
+      // Generate test boilerplate using a subgenerator.
+      this.composeWith(require.resolve('../test'), {
```
Comment (zhawkins) [inline generators/element/index.js:353]: In this file we're moving the "test" specific bits of the generator to their own sub-generator so they can be called as a standalone generator.

File: package.json [hunk 4]
```diff
@@ -22,6 +22,9 @@
     "start": "spandx",
     "lerna": "lerna",
     "test": "node scripts/test.js",
+    "test:watch": "npm config set group=default && web-test-runner --node-resolve --playwright --watch --group=$npm_config_group",
```
Comment (zhawkins) [inline package.json:25]: Using `$npm_config_group` let's us pass commands via arguments like "npm run test:watch --group with-vue".

In this case `--group` maps to `$npm_config_group` which means `$npm_config_group == with-vue`

We use `npm config set group=default` to set a default that we can override so `$npm_config_group == default` if no arguments are passed in.

More here: https://docs.npmjs.com/cli/v7/using-npm/config#environment-variables

File: elements/pfe-select/src/pfe-select.js (hunk 2 above)
Comment (castastrophe) [reply elements/pfe-select/src/pfe-select.js:83]:
```suggestion
    if (this._input) this._input.removeEventListener("input", this._inputChanged);
```
What would you think of this for a slight readability improvement

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -73,13 +70,13 @@ jobs:
 
       - name: Capture hash value
         id: hash
-        run: echo '::set-output name=HASH::${{ runner.os }}-npm-${{ matrix.node }}-${{ hashFiles('**/package-lock.json') }}'
+        run: echo '::set-output name=HASH::${{ runner.os }}-npm-node12-${{ hashFiles('**/package-lock.json') }}'
```
Comment (castastrophe) [inline .github/workflows/tests.yml:73]: Oh hmm, I'm not sure I dig this update.  It's a lot more places to touch to bump the version.  Can we revert and just bump the matrix version to 12?

File: .github/workflows/tests.yml [hunk 6]
```diff
@@ -156,6 +151,57 @@ jobs:
       - name: Run tests
         run: npm test -- --nobuild
 
+  # Test command can run concurrent with e2e and wct so it downloads the
+  # compiled assets from the build and uses those instead of reinstalling.
+  test-wtr:
```
Comment (castastrophe) [inline .github/workflows/tests.yml:156]: Looks great!

File: README.md [hunk 7]
```diff
@@ -62,6 +62,8 @@ npm run dev [component-name(s)]
 
 ### Testing
 
+#### 🕸 Legacy tests ([Web Component Tester](https://github.com/Polymer/web-component-tester))
```
Comment (castastrophe) [inline README.md:65]: 😆 Love the emojis

Comment (castastrophe) [reply README.md:65]: Do you want to bump this below the new tests?

File: elements/pfe-accordion/test/old-test/pfe-accordion_vue_test.html [hunk 8]
```diff
@@ -5,7 +5,7 @@
     <meta name="viewport" content="width=device-width, minimum-scale=1.0, initial-scale=1.0, user-scalable=yes">
     <script src="/components/@webcomponents/webcomponentsjs/webcomponents-bundle.js"></script>
```
Comment (castastrophe) [inline elements/pfe-accordion/test/old-test/pfe-accordion_vue_test.html:6]: You might get the polyfill migration to CDN by updating with the latest from master.

File: elements/pfe-select/package.json [hunk 9]
```diff
@@ -19,7 +19,7 @@
     "build": "../../node_modules/.bin/gulp && ../../node_modules/.bin/prettier --ignore-path ../../.prettierignore --write '**/*.{js,json}'",
     "dev": "../../node_modules/.bin/gulp dev",
     "watch": "../../node_modules/.bin/gulp watch",
-    "test": "../../node_modules/.bin/wct --configFile ../../wct.conf.json elements/pfe-select/test/"
+    "test": "echo \"👉 Run tests from the project root with 'npm run test:watch -- elements/*/test/pfe-select.spec.js'\""
```
Comment (castastrophe) [inline elements/pfe-select/package.json:22]: cc @kylebuch8 for your thoughts on this update ^

File: elements/pfe-select/test/pfe-select.spec.js [hunk 10]
```diff
@@ -0,0 +1,180 @@
+import { assert, expect, oneEvent, elementUpdated } from '@open-wc/testing/index-no-side-effects.js';
+import { spy } from 'sinon';
```
Comment (castastrophe) [inline elements/pfe-select/test/pfe-select.spec.js:2]: Nice, I love sinon

File: generators/test/index.js [hunk 11]
```diff
@@ -0,0 +1,101 @@
+const Generator = require("yeoman-generator");
+const _ = require("lodash");
+const chalk = require("chalk");
+const fg = require("fast-glob");
```
Comment (castastrophe) [inline generators/test/index.js:4]: We have glob installed as a dependency - where is fast-glob?  Should we migrate to use glob for now and update them all to use fast-glob?

File: package.json (hunk 4 above)
Comment (castastrophe) [reply package.json:25]: Love this approach; stealing this to use everywhere :P

File: package.json [hunk 12]
```diff
@@ -22,6 +22,9 @@
     "start": "npm config set quiet=true; eleventy --serve",
     "lerna": "lerna",
     "test": "node scripts/test.js",
+    "test:watch": "npm config set group=default && web-test-runner --node-resolve --playwright --watch --group=$npm_config_group",
+    "test:build:watch": "npm run build && npm run test:watch",
```
Comment (castastrophe) [inline package.json:26]:
```suggestion
    "test:build:watch": "npm run build && npm run test:watch",
```

I'm wondering how reusable this is since most often we're testing a single component and this doesn't allow us to pass in components.  Can we make the test command a script that accepts inputs so we can pass in a build flag or read in the CI env variable?

File: package.json [hunk 13]
```diff
@@ -53,13 +55,15 @@
   },
   "husky": {
     "hooks": {
-      "pre-commit": "npm run test-suite-inject > /dev/null && pretty-quick --staged",
+      "pre-commit": "pretty-quick --staged",
```
Comment (castastrophe) [inline package.json:58]: Good catch!

**Review by castastrophe — CHANGES_REQUESTED** [review]
Looks really good.  I think it would be much more robust to build out a test script though instead of trying to squeeze all the functionality into the package scripts.

File: elements/pfe-select/src/pfe-select.js (hunk 2 above)
Comment (zhawkins) [reply elements/pfe-select/src/pfe-select.js:83]: I'm not a fan of one line `if()` statements but I don't feel that strongly. If that's how it is elsewhere in the code base I'm cool changing it 😎

File: .github/workflows/tests.yml (hunk 5 above)
Comment (zhawkins) [reply .github/workflows/tests.yml:73]: ya sure

File: README.md (hunk 7 above)
Comment (zhawkins) [reply README.md:65]: Ah yes let's do that.

File: elements/pfe-select/src/pfe-select.js (hunk 2 above)
Comment (castastrophe) [reply elements/pfe-select/src/pfe-select.js:83]: I'm cool if you want to wrap it, just thinking the if is easier to understand than the amper syntax

Comment (zhawkins) [reply elements/pfe-select/src/pfe-select.js]: Done!

File: .github/workflows/tests.yml (hunk 5 above)
Comment (zhawkins) [reply .github/workflows/tests.yml]: Done!

File: elements/pfe-select/package.json (hunk 9 above)
Comment (zhawkins) [reply elements/pfe-select/package.json]: The long term goal would be for individual elements to have the right dependencies to run their own tests.

File: generators/test/index.js (hunk 11 above)
Comment (zhawkins) [reply generators/test/index.js]: Ah, my fault. I used that over in WebRH so I used it here. I've since switched it to regular `glob`. I don't think what we're doing would see a distinguishable difference with fast-glob.

File: package.json (hunk 12 above)
Comment (zhawkins) [reply package.json]: Updated per our discussion to support one or many individual components.

**Comment by zhawkins** [comment]
> Looks really good. I think it would be much more robust to build out a test script though instead of trying to squeeze all the functionality into the package scripts.

Updated!

File: README.md [hunk 14]
```diff
@@ -62,6 +62,47 @@ npm run dev [component-name(s)]
...
+#### ✨ New tests ([Web Test Runner](https://modern-web.dev/docs/test-runner/overview/))
+
+```shell
+# Run all tests in watch mode.
+npm run test:watch
+
+# Run a single test in watch mode.
+npm run test:watch --element="pfe-select"
+# Or multiple:
+npm run test:watch --element="{pfe-select,pfe-card}"
```
Comment (zhawkins) [inline README.md:74]: I ended up having to include the `{}` in the parameter. 🙃 

TLDR: `{pfe-select}` won't work. It needs at minimum two strings for the {} to work.

**Review by castastrophe — APPROVED** [review]

File: CHANGELOG-1.x.md [hunk 15]
```diff
@@ -1,6 +1,8 @@
 # 1.6.1 (2021)
```
Comment (castastrophe) [inline CHANGELOG-1.x.md:1]:
```suggestion
# 1.7.0 (2021)
```

File: CHANGELOG-1.x.md [hunk 16]
```diff
@@ -1,6 +1,8 @@
...
 
 - [](https://github.com/patternfly/patternfly-elements/commit/) fix: Updating PFElement.log (aka this.log) so it logs arrays and objects, instead of converting them to strings
```
Comment (castastrophe) [inline CHANGELOG-1.x.md:3]:
```suggestion
- [20cb906](https://github.com/patternfly/patternfly-elements/commit/20cb906b5bf3d9340097170d6be7bdacbb4b12a5) fix: Updating PFElement.log (aka this.log) so it logs arrays and objects, instead of converting them to strings
```
